internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,382 @@
+	if err != nil {
+		return err
+	}
+	// Load a clean X509 certificate from the DER encoded bytes.
+	certificate, err := x509.ParseCertificate(derBytes)
+	if err != nil {
+		return err
+	}
+	m.tlsCert = &tls.Certificate{
+		PrivateKey: privateKey,
+		// Leaf first, then CA cert.
+		Certificate: [][]byte{certificate.Raw, m.idCert.Raw},
+		Leaf:        certificate,
+	}
+	return nil
+}
+
+// createDir verifies the directory layout and permissions, and creates
+// the leaf element of the TLS driectory if it does not yet exist.
+func (m *TLSManager) createDir() error {
+	exists, err := pathExists(m.tlsDir)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return expectPermission(m.tlsDir, 0o700)
+	}
+
+	// Create the leaf directory node with 0o700 permissions.
+	return os.Mkdir(m.tlsDir, 0o700)
+}
+
+// getIDCert verifies that the identity certificate exists, and that the
+// certificate matches the supplied identity key signer. If the identity key
+// has changed, or the certificate has expired, the identity certificate is
+// replaced with a new identity certificate signed with the identity key.
+func (m *TLSManager) getIDCert() error {
+	idPath := filepath.Join(m.tlsDir, idCertFile)
+	exists, err := pathExists(idPath)
+	if err != nil {
+		return err
+	}
+
+	// Load the identity certificate it not yet loaded.
+	if exists && m.idCert == nil {
+		m.idCert, err = loadIDCert(idPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Can we use the existing identity certificate?
+	if m.idCert != nil {
+		isDerived := isCertDerived(m.idCert, m.signer)
+		isActive := isCertActive(m.idCert)
+		if isDerived && isActive {
+			// Existing identity certificate is valid.
+			return nil
+		}
+	}
+
+	// If we get here a new identity certificate must be created.
+	m.idCert, err = createIDCert(m.signer)
+	if err != nil {
+		return err
+	}
+	err = saveIDCert(idPath, m.idCert)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// loadIDCert loads the x509 identity certificate, which contains
+// the identity public key. We only allow a single certificate
+// block in the PEM file.
+func loadIDCert(path string) (*x509.Certificate, error) {
+	err := expectPermission(path, 0o600)
+	if err != nil {
+		return nil, err
+	}
+
+	var cert *x509.Certificate
+	pemData, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	blockCount := 0
+	for {
+		var block *pem.Block
+		block, pemData = pem.Decode(pemData)
+		if block == nil {
+			if blockCount == 1 {
+				break
+			}
+			return nil, fmt.Errorf("empty PEM file")
+		}
+		switch block.Type {
+		case "CERTIFICATE":
+			// We only support a single certificate block.
+			if blockCount == 1 {
+				return nil, fmt.Errorf("unexpected PEM block")
+			}
+			cert, err = x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return nil, err
+			}
+			blockCount += 1
+		default:
+			return nil, fmt.Errorf("unexpected PEM block")
```

Comment:
nit: as in the other PR, how about we ignore unrecognized block types, and make sure to fail if no `CERTIFICATE` block was found?

Reply:
Fixed (matching first PR). Resolving.